neuraxle/base.py

```diff
@@ -561,32 +919,93 @@ def fit_transform(self, data_inputs, expected_outputs=None) -> ('BaseStep', Any)
 
     @abstractmethod
     def fit(self, data_inputs, expected_outputs=None) -> 'BaseStep':
+        """
+        Fit step with the given data inputs, and expected outputs.
+
+        :param data_inputs: data inputs
+        :param expected_outputs: expected outputs to fit on
+        :return: fitted self
+        :rtype: BaseStep
+        """
         raise NotImplementedError("TODO: Implement this method in {}, or have this class inherit from the NonFittableMixin.".format(self.__class__.__name__))
 
     @abstractmethod
     def transform(self, data_inputs):
+        """
+        Transform given data inputs.
+
+        :param data_inputs: data inputs
+        :return: transformed data inputs
+        :rtype: Any
+        """
         raise NotImplementedError("TODO: Implement this method in {}, or have this class inherit from the NonTransformableMixin.".format(self.__class__.__name__))
 
     def inverse_transform(self, processed_outputs):
+        """
+        Inverse Transform the given transformed data inputs.
+
+        :func:`~neuraxle.base.BaseStep.mutate` or :func:`~neuraxle.base.BaseStep.reverse` can be called to change the default transform behavior :
+        ```
+            class MultiplyBy(NonFittableMixin, BaseStep):
+                def transform(self, data_inputs):
+                    return data_inputs * self.hyperparams['multiply_by']
+
+                def inverse_transform(self, processed_outputs):
+                    return data_inputs / self.hyperparams['multiply_by']
+
+            p = Pipeline([MultiplyBy()])
+
+            p = p.mutate(new_method="inverse_transform", method_to_assign_to="transform")
+            p.transform(processed_outputs) # this will call inverse transform method instead
+
+            # is equivalent to :
+
+            p = reverse(p)
+            p.transform(processed_outputs) # this will call inverse transform method instead
+        ```
+
+        :param processed_outputs: processed data inputs
+        :return: inverse transformed processed outputs
+        :rtype: Any
+        """
         raise NotImplementedError("TODO: Implement this method in {}.".format(self.__class__.__name__))
 
     def predict(self, data_input):
+        """
+        Predict data input expected output using transform method.
+
+        :param data_input: data input to predict
+        :return: prediction
+        :rtype: Any
+        """
         return self.transform(data_input)
 
     def should_save(self) -> bool:
         """
         Returns true if the step should be saved.
 
-        :return: bool
+        :return: if the step should be saved
```

Comment:
What is the criteria that says a step should be saved or not? I'd make this docstring much bigger to give example of contexts in which its used.